### PR TITLE
feat: supports reading config values from CLI

### PR DESCRIPTION
**Pull request by anthonykimani**
This PR works on issue #604. 

I using the config-rs library to combine the configs in the `main.rs` file 
image

I also set default values in the `settings.rs` file 

image

image

### CODE REVIEW

**Comment by anthonykimani** [comment]
you can the server with various options:

1. Using defaults - `cargo run`
2. Overriding Specific Settings - `cargo run -- --port 8080 --tls-enabled false --log-level INFO`
3. Using Environment Variables - `NOTARY_SERVER_SERVER__PORT=9000 cargo run`

File: crates/notary/server/src/settings.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    #[serde(flatten)]
+    pub config: NotaryServerProperties,
+}
+
+impl Settings {
+    pub fn new(cli_fields: &CliFields) -> Result<Self, ConfigError> {
+        let mut builder = Config::builder();
+
+        // Add default values
+        builder = builder
```
Comment (yuroitaki) [inline crates/notary/server/src/settings.rs:18]: can we change these to use the config file as the default instead, so that we have these default values only in one location

File: crates/notary/server/src/settings.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        }
+        if let Some(log_level) = &cli_fields.log_level {
+            builder = builder.set_override("logging.level", log_level.clone())?;
+        }
+
+        let config = builder.build()?;
+        let settings: Settings = config.try_deserialize()?;
+
+        // Validate file existence
+        Self::validate_file_exists(&settings.config.tls.private_key_pem_path, "TLS private key")?;
```
Comment (yuroitaki) [inline crates/notary/server/src/settings.rs:62]: we don't have to do these validations as we already handle these in `server.rs`

File: crates/notary/server/src/settings.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        let config_path = Path::new(&cli_fields.config_file);
+        if config_path.exists() {
+            info!("Loading configuration from: {}", cli_fields.config_file);
+            builder = builder.add_source(File::from(config_path));
+        } else {
+            warn!("Config file not found: {}. Using defaults and overrides.", cli_fields.config_file);
+        }
+
+        // Add environment variables
+        builder = builder.add_source(Environment::with_prefix("NOTARY_SERVER").separator("__"));
```
Comment (yuroitaki) [inline crates/notary/server/src/settings.rs:45]: question: so env var will override config file; and CLI will override both of them; is that right?

**Review by yuroitaki — CHANGES_REQUESTED** [review]
Amazing stuff! That was really fast since I just put up the issue yesterday, thanks a lot! Just a few comments :)

File: .idea/.gitignore
Comment (yuroitaki) [inline .idea/.gitignore:1]: please remove this .idea folder :)

File: Cargo.toml [hunk 4]
```diff
@@ -87,6 +87,7 @@ bincode = { version = "1.3" }
 blake3 = { version = "1.5" }
 bytes = { version = "1.4" }
 chrono = { version = "0.4" }
+config = { version = "0.14", features = ["yaml"] }
```
Comment (yuroitaki) [inline Cargo.toml:90]: let's only put this in notary server's Cargo.toml since its the only mod using this crate now

File: crates/notary/server/src/settings.rs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        let mut builder = Config::builder();
+
+        // Add the config file if it exists
+        let config_path = Path::new(&cli_fields.config_file);
+
+        if config_path.exists() {
+            info!("Loading configuration from: {}", cli_fields.config_file);
+            builder = builder.add_source(File::from(config_path));
+        } else {
+            warn!("Config file not found: {}. Using defaults and overrides.", cli_fields.config_file);
```
Comment (yuroitaki) [inline crates/notary/server/src/settings.rs:24]: should throw an error if config file is not found

File: crates/notary/server/src/main.rs [hunk 6]
```diff
@@ -1,25 +1,41 @@
...
+    println!("Loaded settings: {:#?}", settings);
 
     // Set up tracing for logging
-    init_tracing(&config).map_err(|err| eyre!("Failed to set up tracing: {err}"))?;
+    init_tracing(&settings.config)
+        .map_err(|err| eyre!("Failed to set up tracing: {err}"))?;
 
-    debug!(?config, "Server config loaded");
+    debug!(?settings, "Server settings loaded");
+    info!("Server port: {}", settings.config.server.port);
```
Comment (yuroitaki) [inline crates/notary/server/src/main.rs:33]: dont need these info logs as its a lil too verbose

File: crates/notary/server/src/main.rs [hunk 7]
```diff
@@ -1,25 +1,41 @@
...
 async fn main() -> Result<(), NotaryServerError> {
-    // Load command line arguments which contains the config file location
+    // Load command line arguments
     let cli_fields: CliFields = CliFields::from_args();
-    let config: NotaryServerProperties = parse_config_file(&cli_fields.config_file)?;
+
+    let settings = match Settings::new(&cli_fields) {
+        Ok(s) => s,
+        Err(e) => {
+            error!("Failed to load settings: {:?}. Check that all required fields are provided in the configuration.", e);
```
Comment (yuroitaki) [inline crates/notary/server/src/main.rs:17]: the previous code was better and more succinct, curious why did you change into this?

```rust
let settings = Settings::new(&cli_fields)
        .map_err(|err| eyre!("Failed to load settings: {}", err))?;
```

File: crates/notary/server/src/main.rs [hunk 8]
```diff
@@ -1,25 +1,41 @@
...
+            return Err(eyre!(
+            "Failed to load settings: {:?}. Ensure required fields (like `server.port`, `tls.enabled`, etc.) are present.",
+            e
+        ).into());
+        }
+    };
+
+    // Print the entire configuration for debugging
+    println!("Loaded settings: {:#?}", settings);
```
Comment (yuroitaki) [inline crates/notary/server/src/main.rs:26]: we don't use println!, and its also too verbose since you already have L32

File: crates/notary/server/src/domain/cli.rs [hunk 9]
```diff
@@ -5,6 +5,15 @@ use structopt::StructOpt;
 #[structopt(name = "Notary Server")]
 pub struct CliFields {
     /// Configuration file location
-    #[structopt(long, default_value = "./config/config.yaml")]
+    #[structopt(long, default_value = "./server/config/config.yaml")]
```
Comment (yuroitaki) [inline crates/notary/server/src/domain/cli.rs:8]: this is incorrect, since the server is started inside its directory

File: crates/notary/server/config/config.yaml [hunk 10]
```diff
@@ -2,7 +2,7 @@ server:
   name: "notary-server"
   host: "0.0.0.0"
   port: 7047
-  html-info: |
+  html_info: |
```
Comment (yuroitaki) [inline crates/notary/server/config/config.yaml:5]: can we retain kebab-case for the config file?

File: crates/notary/server/config/config.yaml [hunk 11]
```diff
@@ -12,21 +12,21 @@ server:
...
-  max-sent-data: 4096
-  max-recv-data: 16384
+  max_sent_data: 4096
+  max_recv_data: 16384
 
 tls:
   enabled: true
-  private-key-pem-path: "./fixture/tls/notary.key"
-  certificate-pem-path: "./fixture/tls/notary.crt"
+  private_key_pem_path: "../fixture/tls/notary.key"
```
Comment (yuroitaki) [inline crates/notary/server/config/config.yaml:20]: the path was correct, why did you change it?

File: crates/notary/server/src/settings.rs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+            builder = builder.set_override("tls.enabled", tls_enabled)?;
+        }
+        if let Some(log_level) = &cli_fields.log_level {
+            builder = builder.set_override("logging.level", log_level.clone())?;
+        }
+
+        let config = builder.build()?;
+
+        // Log the entire configuration for debugging
+        debug!("Loaded configuration: {:#?}", config);
```
Comment (yuroitaki) [inline crates/notary/server/src/settings.rs:44]: dont need another log here

**Review by yuroitaki — CHANGES_REQUESTED** [review]

**Comment by anthonykimani** [comment]
Sorry @yuroitaki I meant to refactor the commit, I've included the changes you've suggested :) . This commit will load the config.yaml and can also override with `cargo run -- --port 8080 --tls-enabled false --log-level INFO` .

File: crates/notary/server/config/config.yaml (hunk 11 above)
Comment (anthonykimani) [reply crates/notary/server/config/config.yaml]: on my end the path didn't load

File: crates/notary/server/src/settings.rs (hunk 3 above)
Comment (yuroitaki) [reply crates/notary/server/src/settings.rs]: @anthonykimani can u help to address this question?

File: crates/notary/server/src/main.rs [hunk 13]
```diff
@@ -1,25 +1,24 @@
...
+    // Load command line arguments
     let cli_fields: CliFields = CliFields::from_args();
-    let config: NotaryServerProperties = parse_config_file(&cli_fields.config_file)?;
 
-    // Set up tracing for logging
-    init_tracing(&config).map_err(|err| eyre!("Failed to set up tracing: {err}"))?;
+    let settings = Settings::new(&cli_fields)
+        .map_err(|err| eyre!("Failed to load settings: {}", err))?;
 
-    debug!(?config, "Server config loaded");
```
Comment (yuroitaki) [inline crates/notary/server/src/main.rs:19]: lets keep one debug log instead

File: crates/notary/server/src/settings.rs [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+pub struct Settings {
+    #[serde(flatten)]
+    pub config: NotaryServerProperties,
+}
+
+impl Settings {
+    pub fn new(cli_fields: &CliFields) -> Result<Self, ConfigError> {
+        let mut base_path = env::current_dir().expect("Failed to determine the current directory");
+        base_path.pop();
+        let configuration_path = base_path.join("config").join("config.yaml");
```
Comment (yuroitaki) [inline crates/notary/server/src/settings.rs:16]: this path should be from `&cli_fields.config_file` as we want to define it there only and not hardcoded here

**Review by yuroitaki — CHANGES_REQUESTED** [review]
Almost there!

File: crates/notary/server/config/config.yaml (hunk 11 above)
Comment (yuroitaki) [reply crates/notary/server/config/config.yaml]: i cloned ur branch and did the following; and it worked — are you running from `crate/notary/server`?

image
image

File: crates/notary/server/src/settings.rs (hunk 3 above)
Comment (anthonykimani) [reply crates/notary/server/src/settings.rs]: yes the CLI arguments will hold the highest priority, followed with the env var and config file last : )

File: crates/notary/server/src/main.rs (hunk 13 above)
Comment (anthonykimani) [reply crates/notary/server/src/main.rs:19]: awesome

File: crates/notary/server/src/settings.rs (hunk 14 above)
Comment (anthonykimani) [reply crates/notary/server/src/settings.rs:16]: very true, I've updated this to use the `cli_fields`

File: crates/notary/server/src/settings.rs [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+impl Settings {
+    pub fn new(cli_fields: &CliFields) -> Result<Self, ConfigError> {
+        let config_path = Path::new(&cli_fields.config_file);
+
+        let mut builder = Config::builder()
+            // Load base configuration
+            .add_source(File::from(config_path))
+            // Add in settings from environment variables (with a prefix of NOTARY_SERVER and '__' as separator)
+            .add_source(Environment::with_prefix("NOTARY_SERVER").prefix_separator("_").separator("__"));
```
Comment (yuroitaki) [inline crates/notary/server/src/settings.rs:20]: `prefix_separator` should be "__" and `separator` should be "_"; so that we have `NOTARY_SERVER__TLS_ENABLED`

File: crates/notary/server/src/settings.rs [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+            builder = builder.set_override("logging.level", log_level.clone())?;
+        }
+
+        let config = builder.build()?;
+
+        let settings: Settings = config.try_deserialize()?;
+
+        Ok(settings)
+    }
+}
```
Comment (yuroitaki) [inline crates/notary/server/src/settings.rs:39]:
```suggestion
}

```

**Review by yuroitaki — CHANGES_REQUESTED** [review]

File: crates/notary/server/config/config.yaml [hunk 17]
```diff
@@ -29,4 +29,4 @@ logging:
 
 authorization:
   enabled: false
-  whitelist-csv-path: "./fixture/auth/whitelist.csv"
+  whitelist-csv-path: "./fixture/auth/whitelist.csv"
```
[the remark this answers is not in the document]
Comment (yuroitaki) [reply crates/notary/server/config/config.yaml:32]: seems like you still dont have a newline

File: crates/notary/server/src/settings.rs [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+impl Settings {
+    pub fn new(cli_fields: &CliFields) -> Result<Self, ConfigError> {
+        let config_path = Path::new(&cli_fields.config_file);
+
+        let mut builder = Config::builder()
+            // Load base configuration
+            .add_source(File::from(config_path))
+            // Add in settings from environment variables (with a prefix of NOTARY_SERVER and '__' as separator)
+            .add_source(Environment::with_prefix("NOTARY_SERVER").try_parsing(true).prefix_separator("__").separator("_"));
```
Comment (yuroitaki) [inline crates/notary/server/src/settings.rs:20]: i just realised we have to use "__" for separator as well; so that things like notarization.max-sent-data can work.

Comment (yuroitaki) [reply crates/notary/server/src/settings.rs:20]: Also on this, I am sorry but I understand now why you changed the config.yaml to use snake_case before this, cause else env var won't work with those kebab-case config field..

I was hesitant to change the casing back then since it means we have to change that in our server as well, but I guess that's inevitable now. Can you make that change again 🙏

**Review by yuroitaki — CHANGES_REQUESTED** [review]

File: crates/notary/server/src/settings.rs (hunk 18 above)
Comment (anthonykimani) [reply crates/notary/server/src/settings.rs:20]: no worries, let me make the necessary changes.

File: crates/notary/server/src/settings.rs (hunk 16 above)
Comment (anthonykimani) [reply crates/notary/server/src/settings.rs:39]: oh I understand now, we need to have an extra line at the end for uniformity

File: crates/notary/server/src/domain/cli.rs [hunk 19]
```diff
@@ -7,4 +7,13 @@ pub struct CliFields {
     /// Configuration file location
     #[structopt(long, default_value = "./config/config.yaml")]
     pub config_file: String,
+
+    #[structopt(long)]
```
Comment (yuroitaki) [inline crates/notary/server/src/domain/cli.rs:11]:
```suggestion
     /// Port of notary server
    #[structopt(long)]
```

File: crates/notary/server/src/domain/cli.rs [hunk 20]
```diff
@@ -7,4 +7,13 @@ pub struct CliFields {
...
+    pub port: Option<u16>,
+
+    #[structopt(long)]
```
Comment (yuroitaki) [inline crates/notary/server/src/domain/cli.rs:14]:
```suggestion
     /// Flag to turn on/off TLS when connecting to prover
    #[structopt(long)]
```

File: crates/notary/server/src/domain/cli.rs [hunk 21]
```diff
@@ -7,4 +7,13 @@ pub struct CliFields {
...
+    pub tls_enabled: Option<bool>,
+
+    #[structopt(long)]
```
Comment (yuroitaki) [inline crates/notary/server/src/domain/cli.rs:17]:
```suggestion
     /// Level of logging
    #[structopt(long)]
```

**Review by yuroitaki — CHANGES_REQUESTED** [review]
Code is perfect now 🥳 just need to add some documentations — can you add a new section just below [here](https://github.com/tlsnotary/tlsn/tree/dev/crates/notary/server#using-different-setting-files-with-docker) named "Configuration" explaining how people can also use CLI and env var for config, on top of the config file?

**Comment by yuroitaki** [comment]
Also, formatting GH action failed, can you follow the guideline [here](https://github.com/tlsnotary/tlsn/tree/dev/crates/notary/server#using-different-setting-files-with-docker) so that it can pass?

File: crates/notary/server/README.md [hunk 22]
```diff
@@ -66,6 +66,35 @@ docker run --init -p 127.0.0.1:7047:7047 -v <your folder path>:/root/.notary-ser
...
+     max_recv_data: 16384
+    ```
+   
+2. Command-Line Interface (CLI) Arguments    - You can override configuration file settings using CLI arguments when starting the server. CLI arguments take precedence over the config file.
+   - Example:
+   ```shell
+   cargo run -- --port 8080 --tls-enabled false --log-level INFO --max-sent-data 2048
+   ```
+
+3. Environment Variables can also be used to configure the server. They take precedence over both the config file and CLI arguments. The environment variables use the prefix `NOTARY_SERVER__` followed by the configuration path in uppercase, with double underscores as separators.
```
[the remark this answers is not in the document]
Comment (yuroitaki) [reply crates/notary/server/README.md:93]: Also can we be more clear on the double underscores — maybe something like "it's used for nested configuration such as `tls.enabled` in the config file, which will be `TLS__ENABLED` on CLI"

File: crates/notary/server/README.md [hunk 23]
```diff
@@ -66,6 +66,35 @@ docker run --init -p 127.0.0.1:7047:7047 -v <your folder path>:/root/.notary-ser
...
+   - Example:
```
Comment (yuroitaki) [inline crates/notary/server/README.md:94]:
```suggestion
```

File: crates/notary/server/README.md [hunk 24]
```diff
@@ -66,6 +66,35 @@ docker run --init -p 127.0.0.1:7047:7047 -v <your folder path>:/root/.notary-ser
...
+   ```yaml
+   server:
+     name: "notary-server"
+     host: "0.0.0.0"
+     port: 7047
+
+   notarization:
+     max_sent_data: 4096
+     max_recv_data: 16384
```
Comment (yuroitaki) [inline crates/notary/server/README.md:84]:
```suggestion
     max_recv_data: 16384
     ...
```

**Review by yuroitaki — CHANGES_REQUESTED** [review]

File: crates/notary/server/README.md [hunk 25]
```diff
@@ -66,6 +66,35 @@ docker run --init -p 127.0.0.1:7047:7047 -v <your folder path>:/root/.notary-ser
...
+    ```
+   
+2. Command-Line Interface (CLI) Arguments    - You can override configuration file settings using CLI arguments when starting the server. CLI arguments take precedence over the config file.
+   - Example:
+   ```shell
+   cargo run -- --port 8080 --tls-enabled false --log-level INFO --max-sent-data 2048
```
[the remark this answers is not in the document]
Comment (anthonykimani) [reply crates/notary/server/README.md:90]: Yes, nice catch

File: crates/notary/server/README.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (anthonykimani) [reply crates/notary/server/README.md]: yes definitely, your suggestion is easier to understand.

File: crates/notary/server/fixture/auth/whitelist_copied.csv [hunk 26]
```diff
@@ -0,0 +1,4 @@
+"Name","ApiKey","CreatedAt"
```
Comment (yuroitaki) [inline crates/notary/server/fixture/auth/whitelist_copied.csv:1]: You accidentally checked in this I believe

File: crates/notary/server/README.md [hunk 27]
```diff
@@ -82,16 +81,16 @@ docker run --init -p 127.0.0.1:7047:7047 -v <your folder path>:/root/.notary-ser
    notarization:
      max_sent_data: 4096
      max_recv_data: 16384
+   
+   ...
     ```
    
-2. Command-Line Interface (CLI) Arguments    - You can override configuration file settings using CLI arguments when starting the server. CLI arguments take precedence over the config file.
-   - Example:
+2. Command-Line Interface (CLI) Arguments - You can override configuration file settings using CLI arguments when starting the server. They take precedence over both the config file and Environment Variables arguments, e.g.
```
Comment (yuroitaki) [inline crates/notary/server/README.md:88]:
```suggestion
2. Command-Line Interface (CLI) Arguments - You can override *some* configuration file settings using CLI arguments when starting the server. This also takes precedence over the environment variable method below. E.g.
```

File: crates/notary/server/README.md [hunk 28]
```diff
@@ -82,16 +81,16 @@ docker run --init -p 127.0.0.1:7047:7047 -v <your folder path>:/root/.notary-ser
...
    ```shell
-   cargo run -- --port 8080 --tls-enabled false --log-level INFO --max-sent-data 2048
+   cargo run -- --port 8080 --tls-enabled false --log-level INFO
    ```
 
-3. Environment Variables can also be used to configure the server. They take precedence over both the config file and CLI arguments. The environment variables use the prefix `NOTARY_SERVER__` followed by the configuration path in uppercase, with double underscores as separators.
-   - Example: 
+3. Environment Variables can also be used to configure the server and take precedence over the config file. The environment variables use the prefix `NOTARY_SERVER__` followed by the configuration path in uppercase, with double underscores used for nested configuration such that `tls.enabled` in the config file, which will be `TLS__ENABLED` on CLI, e.g.
```
Comment (yuroitaki) [inline crates/notary/server/README.md:93]:
```suggestion
3. Environment Variables - This can be used to configure all the server settings, where it will override the config file. It uses the prefix `NOTARY_SERVER__` followed by the configuration key(s) in uppercase. Double underscores are used in nested configuration keys, e.g. `tls.enabled` in the config file will be `NOTARY_SERVER__TLS__ENABLED`. E.g.
```

File: crates/notary/server/README.md [hunk 29]
```diff
@@ -82,16 +81,16 @@ docker run --init -p 127.0.0.1:7047:7047 -v <your folder path>:/root/.notary-ser
    notarization:
      max_sent_data: 4096
      max_recv_data: 16384
+
```
Comment (yuroitaki) [inline crates/notary/server/README.md:84]:
```suggestion
```

**Review by yuroitaki — CHANGES_REQUESTED** [review]
Final nit 🙏

File: crates/notary/server/fixture/auth/whitelist_copied.csv (hunk 26 above)
Comment (anthonykimani) [reply crates/notary/server/fixture/auth/whitelist_copied.csv:1]: I was running some unit tests, I believe it was automatically generated. removing it now.

**Review by yuroitaki — APPROVED** [review]
@anthonykimani thanks again for the PR!

**Comment by heeckhau** [comment]
Thanks a lot @anthonykimani 🙏 
I will use this very often !
